sdk/dotnet/Pulumi/Serialization/ResourcePackages.cs

```diff
@@ -61,10 +61,14 @@ internal static bool TryGetResourceType(string name, string? version, [NotNullWh
                 from a in LoadReferencedAssemblies()
                 from t in a.GetTypes()
                 where typeof(CustomResource).IsAssignableFrom(t)
-                let attr = t.GetCustomAttribute<ResourceIdentifierAttribute>()
+                let attr = t.GetCustomAttribute<ResourceTypeAttribute>()
                 where attr != null
-                let versionType = (attr.Version, t)
-                group versionType by attr.Token into g
+                let ut = a.GetType($"{a.GetName().Name}.Utilities")
+                let versionProp = ut?.GetProperty("Version", BindingFlags.Static | BindingFlags.Public | BindingFlags.GetProperty)
```

Comment:
Do we want to rely on the internal `Utilities` class being present?

An alternative would be to generate an internal subclass attribute in each provider, and keep this encapsulated like:

In `Pulumi.dll` SDK:

```c#
[AttributeUsage(AttributeTargets.Class)]
public class ResourceTypeAttribute : Attribute
{
    public string Type { get; }
    public string? Version { get; set; }

    public ResourceTypeAttribute(string type) {
        Type = type;
    }
}
```

Generated code:

```c#
internal sealed class AwsResourceTypeAttribute : Pulumi.ResourceTypeAttribute
{
    public AwsResourceTypeAttribute(string type)
        : base(type)
    {
        Version = Utilities.Version;
    }
}

[AwsResourceType("aws:s3/bucket:Bucket")]
public class Bucket : Pulumi.CustomResource
{
}
```

Reply:
I didn't want to create class hierarchies at first but I must admit reliance on Utilities looks worse than I anticipated. I'll explore subclasses tomorrow.

Reply:
Implemented